py4j-python/src/py4j/java_gateway.py

```diff
@@ -678,6 +760,39 @@ def __getattr__(self, name):
         # The name is a method
         return self._methods[name]
 
+    def __dir__(self):
+        self._populate_fields()
+        return list(set(self._methods.keys()) | self._field_names)
+
+    def _populate_fields(self):
+        if not self._fully_populated:
```

Comment:
OK, I will add notes to the places I have.

Now that I am reviewing the code though I am concerned about the ```__dir__``` on the JVMView and what the effect of self._dir_sequence and self._dir_cache being out of step.

In the _current_ use case, ```__dir__``` it is always called from the same thread. But it is possible that someone complicates life in the future, so I'll have a think about it.